### PR TITLE
Tree: Fix bug with determining relative order of concurrently inserted cells

### PR DESCRIPTION
**Pull request by alex-pardes**
## Description

This PR fixes a bug where, when rebasing a branch with multiple changes, we were in some cases unable to determine the relative order a cell created on the source branch and a cell created on the target branch. This is now resolved by using the relative order of the cells' revisions. The cell with the later revision should come before the cell with the earlier revision. This information must now be provided through a RevisionMetadataSource to the top level call to the ChangeRebaser.

### CODE REVIEW

File: experimental/dds/tree2/src/core/rebase/verifyChangeRebaser.ts [hunk 1]
```diff
@@ -4,7 +4,7 @@
  */
 
 import { makeAnonChange } from "./changeRebaser";
-import { ChangeRebaser } from ".";
+import { ChangeRebaser, revisionMetadataSourceFromInfo } from ".";
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/core/rebase/verifyChangeRebaser.ts:7]: That will fix the build error
```suggestion
import { makeAnonChange } from "./changeRebaser";
import { ChangeRebaser, revisionMetadataSourceFromInfo } from ".";
```

File: experimental/dds/tree2/src/shared-tree-core/branch.ts [hunk 2]
```diff
@@ -24,6 +24,8 @@ import {
 } from "../core";
 import { EventEmitter, ISubscribable } from "../events";
 import { fail } from "../util";
+// eslint-disable-next-line import/no-internal-modules
+import { rebaseChangeOverChanges } from "../core/rebase/utils";
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/shared-tree-core/branch.ts:28]: Any reason why this isn't exported by core?

File: experimental/dds/tree2/src/feature-libraries/sequence-field/rebase.ts [hunk 3]
```diff
@@ -1002,16 +1006,26 @@ function compareCellPositions(
...
+
+	if (newRevisionIndex !== undefined) {
+		return Infinity;
+	}
+
+	if (baseRevisionIndex !== undefined) {
+		return -Infinity;
+	}
 
 	// `newMark` points to cells which were emptied before `baseMark` was created.
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/rebase.ts:1030]: THhis whole comment block presumably need updating

Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/sequence-field/rebase.ts:1030]: Actually, do we ever expect `baseRevisionIndex` to be undefined? What would that even mean?

Comment (yann-achard-MS) [reply experimental/dds/tree2/src/feature-libraries/sequence-field/rebase.ts:1030]: Sorry, got confused. It should occur if base is targetting a cell that was created before any of the base changes.

File: experimental/dds/tree2/src/test/feature-libraries/modular-schema/genericFieldKind.spec.ts [hunk 4]
```diff
@@ -121,6 +122,7 @@ const childRebaser = (
 		fakeIdAllocator,
 		crossFieldManager,
 		revisionMetadata,
+		0,
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/modular-schema/genericFieldKind.spec.ts:125]: Is this related to the change of this PR?

```suggestion
		NodeExistenceState.Alive,
```

File: experimental/dds/tree2/src/feature-libraries/modular-schema/modularChangeFamily.ts [hunk 5]
```diff
@@ -749,29 +755,22 @@ export class ModularChangeFamily
...
 	revInfos: readonly RevisionInfo[],
-): RevisionMetadataSource {
-	const getIndex = (revision: RevisionTag): number | undefined => {
-		const index = revInfos.findIndex((revInfo) => revInfo.revision === revision);
-		return index >= 0 ? index : undefined;
-	};
-	const tryGetInfo = (revision: RevisionTag | undefined): RevisionInfo | undefined => {
-		if (revision === undefined) {
-			return undefined;
+	baseRevisions: (RevisionTag | undefined)[],
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/modular-schema/modularChangeFamily.ts:760]: Please add doc comments so it's clear whether `revInfos` should contain the `RevisionInfo` for the base and rebased changes or just the rebased changes.

File: experimental/dds/tree2/src/test/shared-tree/editing.spec.ts [hunk 6]
```diff
@@ -46,6 +46,8 @@ describe("Editing", () => {
 		it("concurrent inserts", () => {
 			const tree1 = makeTreeFromJson([]);
 			const tree2 = tree1.fork();
+
+			// TODO: Could this just be done before the fork?
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/shared-tree/editing.spec.ts:50]: I don't see why not, feel free to fix it.

File: experimental/dds/tree2/src/test/feature-libraries/modular-schema/genericFieldKind.spec.ts (hunk 4 above)
Comment (alex-pardes) [reply experimental/dds/tree2/src/test/feature-libraries/modular-schema/genericFieldKind.spec.ts:125]: I added a numeric parameter at one point and annoyingly the compiler didn't catch this when I removed it. Named parameters might be good when we have this many.

**Review by yann-achard-MS — APPROVED** [review]

File: experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/defaultFieldKinds.spec.ts [hunk 7]
```diff
@@ -275,7 +277,8 @@ describe("defaultFieldKinds", () => {
 					childRebaser,
 					fakeIdAllocator,
 					failCrossFieldManager,
-					defaultRevisionMetadataFromChanges([]),
+					rebaseRevisionMetadataFromInfo([], []),
+					1,
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/default-field-kinds/defaultFieldKinds.spec.ts:281]:
```suggestion
					rebaseRevisionMetadataFromInfo([], []),
```

File: experimental/dds/tree2/src/test/feature-libraries/sequence-field/rebase.spec.ts [hunk 8]
```diff
@@ -937,6 +940,27 @@ describe("SequenceField - Rebase", () => {
 		assert.deepEqual(rebased, expected);
 	});
 
+	it("insert ↷ [delete, insert]", () => {
+		// Because B does not have lineage for A, we should B's insertion's tiebreak policy
+		// and considered the cell it inserts into to be before the cell emptied by A
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/sequence-field/rebase.spec.ts:945]:
```suggestion
		// Because B does not have lineage for A, we should use B's insertion's tiebreak policy
		// and consider the cell it inserts into to be before the cell emptied by A
```

File: experimental/dds/tree2/src/test/feature-libraries/sequence-field/sequenceChangeRebaser.spec.ts [hunk 9]
```diff
@@ -577,9 +580,59 @@ describe("SequenceField - Sandwich Rebasing", () => {
...
+		const rollbackTag1 = mintRevisionTag();
+		const changeA = tagChange([Mark.insert(1, brand(0))], tag1);
+		const inverseA = tagRollbackInverse(invert(changeA), rollbackTag1, tag1);
+
+		const revInfos: RevisionInfo[] = [
+			{ revision: rollbackTag2, rollbackOf: tag2 },
+			{ revision: rollbackTag1, rollbackOf: tag1 },
+			{ revision: tag1 },
+			{ revision: tag2 },
+			{ revision: tag3 },
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/test/feature-libraries/sequence-field/sequenceChangeRebaser.spec.ts:609]: This `{ revision: tag3 }` doesn't seem necessary. It doesn't really bother but I'm curious having it is part of the contract for rebase. If it is I'd like to better understand why, and if not I guess we can remove this line so that we don't risk changing the contract without realizing without the test failing.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/utils.ts [hunk 10]
```diff
@@ -1037,33 +1037,67 @@ function splitDetachEvent(detachEvent: CellId, length: number): CellId {
 }
 
 export function compareLineages(
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/utils.ts:1039]: Would be good to document the function contract here.

File: experimental/dds/tree2/src/feature-libraries/sequence-field/utils.ts [hunk 11]
```diff
@@ -1037,33 +1037,67 @@ function splitDetachEvent(detachEvent: CellId, length: number): CellId {
...
+				return youngerFirst;
 			}
 		}
 	}
 	return 0;
 }
 
+// Returns 1 if cell1 has more lineage and -1 if cell2 has more lineage.
+// Note that this will return zero if both cells are older than the revision metadata and they both
+// have the same lineage, even if they are not the same cell.
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/utils.ts:1082]:
```suggestion
/**
 * Returns 1 if cell1 has more lineage and -1 if cell2 has more lineage.
 * Note that this will return zero if both cells are older than the revision metadata and they both
 * have the same lineage, even if they are not the same cell.
 */
```

File: experimental/dds/tree2/src/feature-libraries/sequence-field/rebase.ts [hunk 12]
```diff
@@ -148,13 +154,14 @@ function rebaseMarkList<TNodeChange>(
 		// Inverse attaches do not contribute to lineage as they are effectively reinstating
 		// an older detach which cells should already have any necessary lineage for.
 		if ((markEmptiesCells(baseMark) || isCellRename(baseMark)) && !isInverseAttach(baseMark)) {
-			const detachId = getOutputCellId(baseMark, baseRevision, metadata);
+			const detachId = getOutputCellId(baseMark, baseRevision, undefined);
```
Comment (yann-achard-MS) [inline experimental/dds/tree2/src/feature-libraries/sequence-field/rebase.ts:157]: Why don't we want to pass the metadata here?

**Review by yann-achard-MS — APPROVED** [review]
